### PR TITLE
LOOP-3845 DIY Sync

### PR DESCRIPTION
**Pull request by ps2**

### CODE REVIEW

File: Loop/Managers/LoopAppManager.swift [hunk 1]
```diff
@@ -210,7 +210,7 @@ class LoopAppManager: NSObject {
         }
         rootNavigationController?.setViewControllers([statusTableViewController], animated: true)
 
-        deviceDataManager.refreshDeviceData()
+        //deviceDataManager.refreshDeviceData()
```
Comment (nhamming) [inline Loop/Managers/LoopAppManager.swift]: [nit] remove?

**Review by nhamming — APPROVED** [review]
LGTM

File: Loop/Managers/DeviceDataManager.swift [hunk 2]
```diff
@@ -683,10 +683,10 @@ extension DeviceDataManager {
         pumpManager.enactBolus(units: units, automatic: automatic) { (error) in
             if let error = error {
                 self.log.error("%{public}@", String(describing: error))
-                
                 switch error {
                 case .uncertainDelivery:
                     // Do not generate notification on uncertain delivery error
```
Comment (darinkrauss) [inline Loop/Managers/DeviceDataManager.swift:688]: [nit] Is this comment still accurate? It looks like the `checkDeliveryUncertaintyState()` call cal post an alert.

File: Loop/Managers/LoopAppManager.swift (hunk 1 above)
Comment (darinkrauss) [reply Loop/Managers/LoopAppManager.swift]: IIRC, if you remove this then the app will remain in red loop for a while after the home screen is initially displayed after onboarding.

File: Loop/Managers/ProfileExpirationAlerter.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        
+        guard let profileExpiration = Bundle.main.profileExpiration, now > profileExpiration - expirationAlertWindow else {
+            return
+        }
+        
+        let timeUntilExpiration = profileExpiration.timeIntervalSince(now)
+        
+        let minimumTimeBetweenAlerts: TimeInterval = timeUntilExpiration > .hours(24) ? .days(2) : .hours(1)
+        
+        if let lastAlert = UserDefaults.appGroup?.lastProfileExpirationAlertDate {
```
Comment (darinkrauss) [inline Loop/Managers/ProfileExpirationAlerter.swift]: [nit] `lastAlert` -> `lastAlertDate` for clarity?

File: Loop/Managers/ProfileExpirationAlerter.swift [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        let formatter = DateComponentsFormatter()
+        formatter.allowedUnits = [.day, .hour]
+        formatter.unitsStyle = .full
+        formatter.zeroFormattingBehavior = .dropLeading
+        formatter.maximumUnitCount = 1
+        let timeUntilExpirationStr = formatter.string(from: timeUntilExpiration)
+
+        let dialog = UIAlertController(
+            title: NSLocalizedString("Profile Expires Soon", comment: "The title for notification of upcoming profile expiration"),
+            message: String(format: NSLocalizedString("Loop will stop working in %@. You will need to update Loop before that, with a new provisioning profile.", comment: "Format string for body for notification of upcoming provisioning profile expiration. (1: amount of time until expiration"), timeUntilExpirationStr!),
```
Comment (darinkrauss) [inline Loop/Managers/ProfileExpirationAlerter.swift]: [nit] I know this should never be an issue for Tidepool Loop, but consider changing instances of "Loop" to the app name.

File: Loop/Managers/ProfileExpirationAlerter.swift [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            preferredStyle: .alert)
+        dialog.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: "Text for ok action on notification of upcoming profile expiration"), style: .default, handler: nil))
+        dialog.addAction(UIAlertAction(title: NSLocalizedString("More Info", comment: "Text for more info action on notification of upcoming profile expiration"), style: .default, handler: { (_) in
+            UIApplication.shared.open(URL(string: "https://loopkit.github.io/loopdocs/build/updating/")!)
```
Comment (darinkrauss) [inline Loop/Managers/ProfileExpirationAlerter.swift:49]: [nit] Very specific to DIY Loop.

File: LoopCore/NSUserDefaults.swift [hunk 6]
```diff
@@ -180,6 +185,19 @@ extension UserDefaults {
         }
     }
     
+    public var lastProfileExpirationAlertDate: Date? {
+        get {
+            if let rawValue = object(forKey: Key.lastProfileExpirationAlertDate.rawValue) as? Date {
```
Comment (darinkrauss) [inline LoopCore/NSUserDefaults.swift]: Perhaps just `return object(forKey: Key.lastProfileExpirationAlertDate.rawValue) as? Date`? The if/else seems unnecessary.

File: Loop/Managers/LoopAppManager.swift [hunk 7]
```diff
@@ -230,6 +230,9 @@ class LoopAppManager: NSObject {
     // MARK: - Life Cycle
 
     func didBecomeActive() {
+        if let rootViewController = rootViewController {
+            ProfileExpirationAlerter.alertIfNeeded(viewControllerToPresentFrom: rootViewController)
```
Comment (darinkrauss) [inline Loop/Managers/LoopAppManager.swift:234]: How does this interact with onboarding if onboarding is still displayed (yes, unlikely)?

**Review by darinkrauss — COMMENTED** [review]
A few comments/nits. I think the commented out `deviceDataManager.refreshDeviceData` in `LoopAppManager` needs to be looked at. Once you've checked that out, ping me and I'll approve.

File: Loop/Managers/LoopAppManager.swift (hunk 1 above)
Comment (ps2) [reply Loop/Managers/LoopAppManager.swift]: Ok, I'll revert this.

File: Loop/Managers/DeviceDataManager.swift (hunk 2 above)
Comment (ps2) [reply Loop/Managers/DeviceDataManager.swift:688]: I reviewed this again, and I think the way it was in Tidepool is correct.

Comment (ps2) [reply Loop/Managers/DeviceDataManager.swift:688]: So I will back out the change below.

File: Loop/Managers/ProfileExpirationAlerter.swift (hunk 5 above)
Comment (ps2) [reply Loop/Managers/ProfileExpirationAlerter.swift:49]: Thanks for pointing this out, but I think that doing anything about it at this point would be over-engineering, as you point out that this should not be an issue for Tidepool Loop.

File: Loop/Managers/LoopAppManager.swift (hunk 7 above)
Comment (ps2) [reply Loop/Managers/LoopAppManager.swift:234]: Just checked; the dialog pops up on top of onboarding, which I think is acceptable; it's just an informative modal dialog that that user has to tap on the ok button to dismiss.

**Pull request merged**